src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp

```diff
@@ -63,72 +64,181 @@ CHIP_ERROR Instance::Init()
 {
-    uint8_t oldPowerSetting = mPowerSetting;
-    mPowerSetting           = powerSetting;
-    if (mPowerSetting != oldPowerSetting)
+    uint32_t oldCookTimeSec = mCookTimeSec;
+    mCookTimeSec            = cookTimeSec;
+    if (mCookTimeSec != oldCookTimeSec)
     {
-        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::PowerSetting::Id);
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::CookTime::Id);
     }
 }
 
 CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
 {
-    ChipLogError(Zcl, "OperationalState: Reading");
+    ChipLogError(Zcl, "Microwave Oven Control: Reading");
     switch (aPath.mAttributeId)
     {
-    case MicrowaveOvenControl::Attributes::CookTime::Id: {
-        ReturnErrorOnFailure(aEncoder.Encode(GetCookTime()));
-    }
-    break;
+    case MicrowaveOvenControl::Attributes::CookTime::Id:
+        ReturnErrorOnFailure(aEncoder.Encode(GetCookTimeSec()));
+        break;
 
-    case MicrowaveOvenControl::Attributes::PowerSetting::Id: {
-        ReturnErrorOnFailure(aEncoder.Encode(GetPowerSetting()));
-    }
-    break;
+    case MicrowaveOvenControl::Attributes::MaxCookTime::Id:
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetMaxCookTimeSec()));
+        break;
 
-    case MicrowaveOvenControl::Attributes::MinPower::Id: {
-        ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetMinPower()));
-    }
-    break;
+    case MicrowaveOvenControl::Attributes::PowerSetting::Id:
+        if (HasFeature(MicrowaveOvenControl::Feature::kPowerAsNumber))
+        {
+            ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetPowerSettingNum()));
+        }
+        else
+        {
+            ChipLogError(Zcl, "Microwave Oven Control: can not get PowerSetting number, feature is not supported");
+            return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+        }
+        break;
 
-    case MicrowaveOvenControl::Attributes::MaxPower::Id: {
-        ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetMaxPower()));
-    }
-    break;
+    case MicrowaveOvenControl::Attributes::MinPower::Id:
+        if (HasFeature(MicrowaveOvenControl::Feature::kPowerAsNumber))
+        {
+            ReturnErrorOnFailure(aEncoder.Encode(
+                HasFeature(MicrowaveOvenControl::Feature::kPowerNumberLimits) ? mDelegate->GetMinPowerNum() : kDefaultMinPowerNum));
+        }
+        else
+        {
+            ChipLogError(Zcl, "Microwave Oven Control: can not get MinPower number, feature is not supported");
+            return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+        }
+        break;
 
-    case MicrowaveOvenControl::Attributes::PowerStep::Id: {
-        ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetPowerStep()));
-    }
-    break;
+    case MicrowaveOvenControl::Attributes::MaxPower::Id:
+        if (HasFeature(MicrowaveOvenControl::Feature::kPowerAsNumber))
```

Comment:
nit: if you invert conditionals, you may get shorter code (with early return on error):

```cpp
case MicrowaveOvenControl::Attributes::MaxPower::Id:
    VerifyOrReturnError(
        HasFeature(MicrowaveOvenControl::Feature::kPowerAsNumber)),
        CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE,
        ChipLogError(Zcl, "Microwave Oven Control: can not get MaxPower number, feature is not supported")
    );
    
    ReturnErrorOnFailure(aEncoder.Encode(
                HasFeature(MicrowaveOvenControl::Feature::kPowerNumberLimits) ? mDelegate->GetMaxPowerNum() : kDefaultMaxPowerNum));
    break;
```

I am not sure we need the log on unsupported feature in the first place - we return the right code according to the spec, not sure why we would log that the cluster is asked for data it does not support.

Reply:
In case developer has not turn off the attribute fields in ZAP file and the feature is not supported for some fields, when the read attributes command is called, the chip-tool (client side) will return FAILURE on these fields, and the device (server side) will also log "Microwave Oven Control: can not get MaxPower number, feature is not supported", the developer will know next step is to turn off the attribute fields in ZAP file.